test/integration/crud/bulk.test.ts

```diff
@@ -104,6 +104,135 @@ describe('Bulk', function () {
           }
         });
       });
+      context('when inserting duplicate values', function () {
+        async function assertFailsWithDuplicateFields(
+          input,
+          isOrdered,
+          indices,
+          expectedInsertedIds
+        ) {
+          try {
+            const db = client.db();
+            const col = db.collection('test');
+            for (let i = 0; i < indices.length; i++) {
+              await col.createIndex(indices[i], { unique: true, sparse: false });
+            }
+            await col.insertMany(input, { ordered: isOrdered });
+            expect.fail('a MongoBulkWriteError must be thrown when write errors are encountered');
+          } catch (error) {
+            expect(error instanceof MongoBulkWriteError).to.equal(true);
+            expect(error.result.insertedCount).to.equal(
+              Object.keys(error.result.insertedIds).length
+            );
+            expect(error.result.insertedIds).to.deep.equal(expectedInsertedIds);
+          }
+        }
+        context('when the insert is ordered', function () {
+          it('contains the correct insertedIds on one duplicate insert', async function () {
```

Comment:
You can leave the tests as is, I just want to share an alternative organizational format, there were many more tests I would encourage us to reformat them as a table to improve the maintainence expirence.

When all your tests share the same functionality, but operate on different inputs, you can succinctly "table"-ify them with a loop:
```js
context('all the tests', () => {
  const table = [
    { title: 'contains the correct insertedIds on one duplicate insert', input: [ { _id: 0, a: 1 }, { _id: 1, a: 1 } ], isOrdered: true, expected: { 0: 0 } },
    // ... 
  ]
  for (const { title, input, isOrdered, expected } of table)
    it(title, async () => assertFailsWithDuplicateFields(input, isOrdered, expected))
})

```

Reply:
This is good to know, I think I'll hold off on it for these tests, though.